FIPS enabled on Build server fails on Run Code Analysis

On May 12, our Azure DevOps extension updated itself and immediately stopped with an error on the "Run Code Analysis" task.

Due to requirements of our project the build server must have FIPS enabled, and it looks like according to the changes made in the Azure DevOps plugin version 4.10.0 that was just released, it added a SHA256Managed algorithm which is not FIPS compliant. (see https://github.com/SonarSource/sonar-scanner-msbuild/commit/5c5e07f46b5d654830e326e55b11dd54a83cb0cc)

```
SonarScanner for MSBuild 4.9
Using the .NET Framework version of the Scanner for MSBuild
Post-processing started.
09:07:42.165  Fetching code coverage report information from TFS...
09:07:42.165  Attempting to locate a test results (.trx) file...
09:07:44.571  Looking for TRX files in: D:\agents\agent2\_work\180\TestResults
09:07:44.571  No test results files found
09:07:45.602  Did not find any binary coverage files in the expected location.
09:07:45.602  Falling back on locating coverage files in the agent temp directory.
09:07:45.602  Searching for coverage files in D:\agents\agent2\_work\_temp

##[error]Unhandled Exception:
##[error]System.InvalidOperationException: This implementation is not part of the Windows Platform FIPS validated cryptographic algorithms.
   at System.Security.Cryptography.SHA256Managed..ctor()
   at SonarScanner.MSBuild.TFS.BuildVNextCoverageSearchFallback.<>c.<FindCoverageFiles>b__5_0(String fullFilePath)
```

--- a/its/pom.xml
+++ b/its/pom.xml
@@ -31,7 +31,7 @@
     <dependency>
       <groupId>org.sonarsource.orchestrator</groupId>
       <artifactId>sonar-orchestrator</artifactId>
-      <version>3.28.0.2379</version>
+      <version>3.29.0.2543</version>
       <scope>test</scope>
     </dependency>
     <dependency>

--- a/its/src/test/java/com/sonar/it/scanner/msbuild/CppTest.java
+++ b/its/src/test/java/com/sonar/it/scanner/msbuild/CppTest.java
@@ -91,7 +91,7 @@ public class CppTest {
       String.format("/p:WindowsTargetPlatformVersion=%s", windowsSdk),
       String.format("/p:PlatformToolset=%s", plateformToolset));
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, projectKey);
     assertThat(result.isSuccess()).isTrue();
     assertThat(result.getLogs()).doesNotContain("Invalid character encountered in file");
 
@@ -137,7 +137,7 @@ public class CppTest {
       String.format("/p:WindowsTargetPlatformVersion=%s", windowsSdk),
       String.format("/p:PlatformToolset=%s", plateformToolset));;
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, projectKey);
     assertThat(result.isSuccess()).isTrue();
     assertThat(result.getLogs()).doesNotContain("Invalid character encountered in file");
 

--- a/its/src/test/java/com/sonar/it/scanner/msbuild/SQLServerTest.java
+++ b/its/src/test/java/com/sonar/it/scanner/msbuild/SQLServerTest.java
@@ -58,7 +58,7 @@ public class SQLServerTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
     assertThat(issues).hasSize(3);

--- a/its/src/test/java/com/sonar/it/scanner/msbuild/ScannerMSBuildTest.java
+++ b/its/src/test/java/com/sonar/it/scanner/msbuild/ScannerMSBuildTest.java
@@ -30,9 +30,7 @@ import java.nio.file.LinkOption;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Arrays;
-import java.util.Collections;
 import java.util.List;
-import java.util.Set;
 import java.util.concurrent.ConcurrentLinkedDeque;
 import java.util.function.Function;
 import java.util.stream.Collectors;
@@ -69,7 +67,6 @@ import org.slf4j.LoggerFactory;
 import org.sonarqube.ws.Components;
 import org.sonarqube.ws.Issues.Issue;
 import org.sonarqube.ws.client.WsClient;
-import org.sonarqube.ws.client.components.SearchRequest;
 import org.sonarqube.ws.client.components.ShowRequest;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -106,7 +103,6 @@ public class ScannerMSBuildTest {
   }
 
 
-
   @Test
   public void testSample() throws Exception {
     ORCHESTRATOR.getServer().restoreProfile(FileLocation.of("projects/ProjectUnderTest/TestQualityProfile.xml"));
@@ -123,7 +119,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertTrue(result.isSuccess());
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
     assertThat(issues).hasSize(2);
@@ -160,7 +156,7 @@ public class ScannerMSBuildTest {
       .setEnvironmentVariable("SONAR_SCANNER_OPTS",
         "-Dhttp.nonProxyHosts= -Dhttp.proxyHost=localhost -Dhttp.proxyPort=" + httpProxyPort + " -Dhttp.proxyUser=" + PROXY_USER + " -Dhttp.proxyPassword=" + PROXY_PASSWORD));
 
-    TestUtils.dumpComponentList(ORCHESTRATOR);
+    TestUtils.dumpComponentList(ORCHESTRATOR, PROJECT_KEY);
     TestUtils.dumpAllIssues(ORCHESTRATOR);
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
@@ -198,7 +194,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
     assertThat(issues).hasSize(2);
@@ -230,7 +226,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertTrue(result.isSuccess());
 
     // Dump debug info
@@ -273,7 +269,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertTrue(result.isSuccess());
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
@@ -309,7 +305,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertTrue(result.isSuccess());
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
     List<String> ruleKeys = issues.stream().map(Issue::getRule).collect(Collectors.toList());
@@ -320,7 +316,7 @@ public class ScannerMSBuildTest {
       "vbnet:S112",
       "vbnet:S3385"));
 
-    if (ORCHESTRATOR.getServer().version().isGreaterThanOrEquals(7,4)) {
+    if (ORCHESTRATOR.getServer().version().isGreaterThanOrEquals(7, 4)) {
       // if external issues are imported, then there should also be some CodeCracker errors.
       assertThat(ruleKeys).containsAll(Arrays.asList(
         "external_roslyn:CC0021",
@@ -350,7 +346,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertTrue(result.isSuccess());
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
@@ -428,7 +424,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertThat(result.isSuccess()).isTrue();
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
@@ -450,7 +446,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
 
     assertThat(result.getLogs()).doesNotContain("File is not under the project directory and cannot currently be analysed by SonarQube");
     assertThat(result.getLogs()).doesNotContain("AssemblyAttributes.cs");
@@ -471,7 +467,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertTrue(result.isSuccess());
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
@@ -483,8 +479,7 @@ public class ScannerMSBuildTest {
       "csharpsquid:S125",
       "csharpsquid:S1134"));
 
-    if (ORCHESTRATOR.getServer().version().isGreaterThanOrEquals(7,4))
-    {
+    if (ORCHESTRATOR.getServer().version().isGreaterThanOrEquals(7, 4)) {
       // if external issues are imported, then there should also be some
       // Wintellect errors.  However, only file-level issues are imported.
       assertThat(ruleKeys).containsAll(Arrays.asList(
@@ -515,7 +510,7 @@ public class ScannerMSBuildTest {
     TestUtils.runNuGet(ORCHESTRATOR, projectDir, "restore");
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild", "/nr:false");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertTrue(result.isSuccess());
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
@@ -543,7 +538,7 @@ public class ScannerMSBuildTest {
     TestUtils.runNuGet(ORCHESTRATOR, projectDir, "restore");
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild", "/nr:false");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, PROJECT_KEY);
     assertTrue(result.isSuccess());
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
@@ -563,7 +558,7 @@ public class ScannerMSBuildTest {
     ORCHESTRATOR.getServer().associateProjectToQualityProfile(folderName, "cs",
       "ProfileForTestCustomRoslyn");
 
-    runBeginBuildAndEndForStandardProject(folderName,"" ,false);
+    runBeginBuildAndEndForStandardProject(folderName, "", false);
 
     List<Issue> issues = TestUtils.allIssues(ORCHESTRATOR);
     assertThat(issues).hasSize(2 + 37 + 1);
@@ -571,14 +566,14 @@ public class ScannerMSBuildTest {
 
   @Test
   public void testCSharpAllFlat() throws IOException {
-    runBeginBuildAndEndForStandardProject("CSharpAllFlat","",  true);
+    runBeginBuildAndEndForStandardProject("CSharpAllFlat", "", true);
 
     assertThat(getComponent("CSharpAllFlat:Common.cs")).isNotNull();
   }
 
   @Test
   public void testCSharpSharedFiles() throws IOException {
-    runBeginBuildAndEndForStandardProject("CSharpSharedFiles", "",true);
+    runBeginBuildAndEndForStandardProject("CSharpSharedFiles", "", true);
 
     assertThat(getComponent("CSharpSharedFiles:Common.cs"))
       .isNotNull();
@@ -592,7 +587,7 @@ public class ScannerMSBuildTest {
 
   @Test
   public void testCSharpSharedProjectType() throws IOException {
-    runBeginBuildAndEndForStandardProject("CSharpSharedProjectType", "",true);
+    runBeginBuildAndEndForStandardProject("CSharpSharedProjectType", "", true);
 
     assertThat(getComponent("CSharpSharedProjectType:SharedProject/TestEventInvoke.cs"))
       .isNotNull();
@@ -606,21 +601,24 @@ public class ScannerMSBuildTest {
 
   @Test
   public void testCSharpSharedFileWithOneProjectWithoutProjectBaseDir() throws IOException {
-    runBeginBuildAndEndForStandardProject("CSharpSharedFileWithOneProject", "ClassLib1",true);
+    runBeginBuildAndEndForStandardProject("CSharpSharedFileWithOneProject", "ClassLib1", true);
 
-    Set<String> componentKeys = newWsClient()
-      .components()
-      .search(new SearchRequest().setLanguage("cs").setQualifiers(Collections.singletonList("FIL")))
-      .getComponentsList()
-      .stream()
-      .map(Components.Component::getKey)
-      .collect(Collectors.toSet());
+    try {
+      Components.ShowWsResponse showComponentResponse = newWsClient()
+        .components()
+        .show(new ShowRequest().setComponent("CSharpSharedFileWithOneProject:Common.cs"));
+    } catch (org.sonarqube.ws.client.HttpException ex) {
+      assertThat(ex.code()).isEqualTo(404);
+    }
 
     // When not using /d:sonar.projectBaseDir the root dir will be set at the level of the project so that the
     // file Common.cs will be outside of the scope and won't be pushed to SQ
-    assertThat(componentKeys).doesNotContain("CSharpSharedFileWithOneProject:Common.cs");
-    String class1ComponentId = TestUtils.hasModules(ORCHESTRATOR) ? "CSharpSharedFileWithOneProject:CSharpSharedFileWithOneProject:D8FEDBA2-D056-42FB-B146-5A409727B65D:Class1.cs" : "CSharpSharedFileWithOneProject:Class1.cs";
-    assertThat(componentKeys).contains(class1ComponentId);
+
+    Components.ShowWsResponse showComponentResponse2 = newWsClient()
+      .components()
+      .show(new ShowRequest().setComponent("CSharpSharedFileWithOneProject:Class1.cs"));
+
+    assertThat(showComponentResponse2.hasComponent()).isTrue();
   }
 
   @Test
@@ -663,7 +661,7 @@ public class ScannerMSBuildTest {
 
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild");
 
-    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    BuildResult result = TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, folderName);
     assertTrue(result.isSuccess());
     assertThat(getComponent(folderName + ":Common.cs"))
       .isNotNull();
@@ -690,10 +688,9 @@ public class ScannerMSBuildTest {
       // the issue is to explicitly set the projectBaseDir to the project directory, as this
       // will take precedence, so then then key for the shared file is what is expected by
       // the tests.
-      if(projectName.isEmpty())
-      {
+      if (projectName.isEmpty()) {
         scanner.addArgument("/d:sonar.projectBaseDir=" + projectDir.toAbsolutePath());
-      }else{
+      } else {
         scanner.addArgument("/d:sonar.projectBaseDir=" + Paths.get(projectDir.toAbsolutePath().toString(), projectName).toString());
       }
 
@@ -701,7 +698,7 @@ public class ScannerMSBuildTest {
 
     ORCHESTRATOR.executeBuild(scanner);
     TestUtils.runMSBuild(ORCHESTRATOR, projectDir, "/t:Rebuild", folderName + ".sln");
-    TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir);
+    TestUtils.executeEndStepAndDumpResults(ORCHESTRATOR, projectDir, folderName);
   }
 
   private static Components.Component getComponent(String componentKey) {
@@ -753,7 +750,7 @@ public class ScannerMSBuildTest {
     HashLoginService l = new HashLoginService();
 
     UserStore userStore = new UserStore();
-    userStore.addUser(username, Credential.getCredential(password), new String[] { "user"});
+    userStore.addUser(username, Credential.getCredential(password), new String[]{"user"});
 
     l.setUserStore(userStore);
 

--- a/its/src/test/java/com/sonar/it/scanner/msbuild/TestUtils.java
+++ b/its/src/test/java/com/sonar/it/scanner/msbuild/TestUtils.java
@@ -49,6 +49,7 @@ import org.sonarqube.ws.Components;
 import org.sonarqube.ws.Issues.Issue;
 import org.sonarqube.ws.Measures;
 import org.sonarqube.ws.client.components.SearchRequest;
+import org.sonarqube.ws.client.components.TreeRequest;
 import org.sonarqube.ws.client.measures.ComponentRequest;
 import org.sonarqube.ws.client.HttpConnector;
 import org.sonarqube.ws.client.WsClient;
@@ -253,11 +254,11 @@ public class TestUtils {
     return msBuildPath;
   }
 
-  static void dumpComponentList(Orchestrator orchestrator)
+  static void dumpComponentList(Orchestrator orchestrator, String projectKey)
   {
     Set<String> componentKeys = newWsClient(orchestrator)
       .components()
-      .search(new SearchRequest().setLanguage("cs").setQualifiers(Collections.singletonList("FIL")))
+      .tree(new TreeRequest().setQualifiers(Collections.singletonList("FIL")).setComponent(projectKey))
       .getComponentsList()
       .stream()
       .map(Components.Component::getKey)
@@ -276,12 +277,12 @@ public class TestUtils {
     }
   }
 
-  static BuildResult executeEndStepAndDumpResults(Orchestrator orchestrator, Path projectDir){
+  static BuildResult executeEndStepAndDumpResults(Orchestrator orchestrator, Path projectDir, String projectKey){
     BuildResult result = orchestrator.executeBuild(TestUtils.newScanner(orchestrator, projectDir)
       .addArgument("end"));
 
     if (result.isSuccess()) {
-      TestUtils.dumpComponentList(orchestrator);
+      TestUtils.dumpComponentList(orchestrator, projectKey);
       TestUtils.dumpAllIssues(orchestrator);
     }
     else

--- a/src/SonarScanner.MSBuild.TFS/IBuildVNextCoverageSearchFallback.cs
+++ b/src/SonarScanner.MSBuild.TFS/IBuildVNextCoverageSearchFallback.cs
@@ -82,7 +82,7 @@ namespace SonarScanner.MSBuild.TFS
             {
                 using (var fileStream = new FileStream(fullFilePath, FileMode.Open))
                 using (var bufferedStream = new BufferedStream(fileStream))
-                using (var sha = new SHA256Managed())
+                using (var sha = new SHA256CryptoServiceProvider())
                 {
                     var contentHash = sha.ComputeHash(bufferedStream);
 

